subsys/bluetooth/host/adv.c

```diff
@@ -1896,18 +1897,37 @@ void bt_hci_le_adv_set_terminated(struct net_buf *buf)
 	struct bt_hci_evt_le_adv_set_terminated *evt;
 	struct bt_le_ext_adv *adv;
 	uint16_t conn_handle;
+#if (CONFIG_BT_ID_MAX > 1) && (CONFIG_BT_EXT_ADV_MAX_ADV_SET > 1)
+	bool was_adv_enabled;
+#endif
 
 	evt = (void *)buf->data;
 	adv = bt_adv_lookup_handle(evt->adv_handle);
 	conn_handle = sys_le16_to_cpu(evt->conn_handle);
 
 	(void)bt_le_lim_adv_cancel_timeout(adv);
+
+	BT_DBG("status 0x%02x adv_handle %u conn_handle 0x%02x num %u",
+	       evt->status, evt->adv_handle, conn_handle,
+	       evt->num_completed_ext_adv_evts);
+
+	if (!adv) {
+		BT_ERR("No valid adv");
+		return;
+	}
+
+#if (CONFIG_BT_ID_MAX > 1) && (CONFIG_BT_EXT_ADV_MAX_ADV_SET > 1)
+	was_adv_enabled = atomic_test_bit(adv->flags, BT_ADV_ENABLED);
+#endif
+
+	atomic_clear_bit(adv->flags, BT_ADV_ENABLED);
+
 #if (CONFIG_BT_ID_MAX > 1) && (CONFIG_BT_EXT_ADV_MAX_ADV_SET > 1)
 	bt_dev.adv_conn_id = adv->id;
 	for (int i = 0; i < ARRAY_SIZE(bt_dev.cached_conn_complete); i++) {
 		if (bt_dev.cached_conn_complete[i].valid &&
 		    bt_dev.cached_conn_complete[i].evt.handle == evt->conn_handle) {
-			if (atomic_test_bit(adv->flags, BT_ADV_ENABLED)) {
+			if (was_adv_enabled) {
```

Comment:
So this change is to make connectable extended advertising persistable? 

Is this fully working? 

In `bluetooth.h` we have 
```c
	/**
	 * @brief Advertise one time.
	 *
	 * Don't try to resume connectable advertising after a connection.
	 * This option is only meaningful when used together with
	 * BT_LE_ADV_OPT_CONNECTABLE. If set the advertising will be stopped
	 * when bt_le_adv_stop() is called or when an incoming (peripheral)
	 * connection happens. If this option is not set the stack will
	 * take care of keeping advertising enabled even as connections
	 * occur.
	 * If Advertising directed or the advertiser was started with
	 * @ref bt_le_ext_adv_start then this behavior is the default behavior
	 * and this flag has no effect.
	 */
	BT_LE_ADV_OPT_ONE_TIME = BIT(1),
```

specifically because extended advertising is not persistable. 

Besides that, we can have multiple connectable advertising sets, and just having one set taking exclusive rights on being connectable like here, perhaps isn't what we want. See the issue I raised in https://github.com/zephyrproject-rtos/zephyr/issues/30699.

Reply:
Actually, this was to fix legacy advertising when using the extended advertising HCI interface. A lot of our tests started failing once we enabled CONFIG_BT_EXT_ADV/CONFIG_BT_CTLR_ADV_EXT while still only using legacy advertising.

But if it is specifically not supported when using extended advertising in the HCI, then I guess we will have to live with it.

Reply:
> HCI does not have the concept of persistable advertising at all - That is a Zephyr feature.
> 
> If you are not using `bt_le_ext_adv_start` directly, persisting advertising should still work, but I'll have to try out the BT shell or something to determine if there's indeed a problem and then do a proper review of this change.

Sorry I was probably a bit unclear. We are using `bt_le_adv_start()` (without BT_LE_ADV_OPT_ONE_TIME) to start advertising and resume advertising after a connection. After turning on the extended advertising support it no longer resumes the advertising though.

My fix does not take this bit into account though:
`	 * If Advertising directed or the advertiser was started with
	 * @ref bt_le_ext_adv_start then this behavior is the default behavior
	 * and this flag has no effect.
`

I'm not sure how we can determine when advertising has been started by a call to `bt_le_ext_adv_start()` vs `bt_le_adv_start()` though.

Reply:
> I'm not sure how we can determine when advertising has been started by a call to bt_le_ext_adv_start() vs bt_le_adv_start() though.

`bt_le_adv_start()` will always use the `bt_dev.adv` advertiser, and I think that the `resume` function will just use that. 

> After turning on the extended advertising support it no longer resumes the advertising though.

OK, so this change is to fix resuming when extended advertising is enabled, but when still using `bt_le_adv_start()` - That makes sense. I'll have to run a few tests locally to see if there's indeed a general issue (none has reported before AFAIK), and it has been working before IIRC.

Reply:
> > I'm not sure how we can determine when advertising has been started by a call to bt_le_ext_adv_start() vs bt_le_adv_start() though.
> 
> `bt_le_adv_start()` will always use the `bt_dev.adv` advertiser, and I think that the `resume` function will just use that.

Ah yes as far as I can tell `bt_le_adv_resume()` will only resume the legacy advertising (and only when relevant). So it should be safe to call it the way this fix does.

Reply:
@Tronil Can you elaborate on why/when the advertising isn't resumed? I just used the shell on `main` with `CONFIG_BT_EXT_ADV=y` and where `bt_le_adv_start` calls `bt_le_adv_start_ext`, and where `bt_le_adv_resume` is called on disconnect and thus calling `bt_le_adv_set_enable` again. 

So to me it seems to work as intended.

Reply:
> @Tronil Can you elaborate on why/when the advertising isn't resumed? I just used the shell on `main` with `CONFIG_BT_EXT_ADV=y` and where `bt_le_adv_start` calls `bt_le_adv_start_ext`, and where `bt_le_adv_resume` is called on disconnect and thus calling `bt_le_adv_set_enable` again.

The problem isn't on disconnect - rather, advertising should be resumed after a connection has been established, so we can establish several connections at the same time. I would expect it to resume after a disconnect since BT_ADV_ENABLED has then long since been cleared - however, that is too late.